Prog/Neighborhood change/4.1_prediction_OTR.R

```diff
@@ -0,0 +1,338 @@
+  predicted_probs <- predict(model, data, type = "response")
+  
+  # Classify observations based on a 0.5 threshold
+  predicted_classes <- ifelse(predicted_probs > 0.5, 1, 0)
+  
+  # Actual classes
+  actual_classes <- data$displacement
+  
+  # Calculate accuracy
+  accuracy <- mean(predicted_classes == actual_classes)
+  
+  # Calculate confusion matrix components
+  true_positives <- sum((predicted_classes == 1) & (actual_classes == 1))
+  true_negatives <- sum((predicted_classes == 0) & (actual_classes == 0))
+  false_positives <- sum((predicted_classes == 1) & (actual_classes == 0))
+  false_negatives <- sum((predicted_classes == 0) & (actual_classes == 1))
+  
+  # Calculate False Positive Rate (FPR) and False Negative Rate (FNR)
+  total_actual_negatives <- sum(actual_classes == 0)
+  total_actual_positives <- sum(actual_classes == 1)
+  
+  FPR <- false_positives / total_actual_negatives
+  FNR <- false_negatives / total_actual_positives
+  
+  # Return the metrics as a named list
+  return(list(accuracy = accuracy, FPR = FPR, FNR = FNR))
+}
+
+# Create an empty data frame to store metrics for each model
+metrics_df <- data.frame(Model = character(), Accuracy = numeric(), FPR = numeric(), FNR = numeric(), stringsAsFactors = FALSE)
+
+# List of models for easier looping
+models <- list(logit = logit, logit2 = logit2, logit3 = logit3, logit4 = logit4, logit5 = logit5, logit6 = logit6, logit7 = logit7) # Add more models here if needed
+
+# Loop over each model, calculate metrics, and store them in the data frame
+for (model_name in names(models)) {
+  metrics <- calculate_metrics(models[[model_name]], predictionmaster1)
+  
+  # Append the metrics to the data frame
+  metrics_df <- rbind(metrics_df, data.frame(
+    Model = model_name,
+    Accuracy = metrics$accuracy,
+    FPR = metrics$FPR,
+    FNR = metrics$FNR
+  ))
+}
+
+# Print the final data frame with accuracy, FPR, and FNR for each model
+print(metrics_df)
+
+
+
+
+
+predicted_probs <- predict(logit3, predictionmaster1, type = "response")
+
+# Classify observations based on a 0.5 threshold
+predicted_classes <- ifelse(predicted_probs > 0.5, 1, 0)
+
+# Actual classes
+actual_classes <- predictionmaster1$displacement
+
+# Calculate accuracy
+accuracy <- mean(predicted_classes == actual_classes)
+
+
+
+#################predict
+
+# Predict probabilities of "am" being 1 (manual transmission)
+predicted_probs <- predict(logit6, predictionmaster2, type = "response")
+print(predicted_probs)
+# Convert probabilities to predicted class (0 or 1) based on a threshold of 0.5
+predicted_class <- ifelse(predicted_probs > 0.5, 1, 0)
+
+# Print predicted classes
+print(predicted_class)
+
+predictionmaster2$predicted_probs <- predicted_probs
+predictionmaster2$predicted_class <- predicted_class
+
+Testresult <- predictionmaster2 %>% 
+  left_join(neighborhoodname, by=c("GEOID")) %>% 
+  select(GEOID, NBH_NAMES,Ward, NAME.y, displacement, predicted_probs,predicted_class )
```

Comment:
I don't know why, but this generated an error. The column looked like it was in the neighborhoodname table. 
```
Error in `select()`:
! Can't select columns that don't exist.
✖ Column `NBH_NAMES` doesn't exist.
```

Reply:
yes, sorry, I might have created some temporary export in the previous step where neighborhood name was already in there, causing two new variables NBH_NAMES.x and NBH_NAMES.y